packager/amp_cache_transform/amp_cache_transform.go

```diff
@@ -0,0 +1,139 @@
+// Parses and interprets the AMP-Cache-Transform header specified at
+// https://github.com/ampproject/amphtml/blob/master/spec/amp-cache-transform.md.
+
+package amp_cache_transform
+
+import (
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-3.3
+type parameterisedIdentifier struct {
+	id     string
+	// In the future, this will include params.
+}
+
+// https://tools.ietf.org/html/rfc7230#appendix-B (OWS = "optional whitespace")
+func discardOWS(reader *strings.Reader) error {
+	for reader.Len() > 0 {
+		c, err := reader.ReadByte()
+		if err != nil {
+			return errors.Wrap(err, "reading OWS")
+		}
+		if c != ' ' && c != '\t' {
+			// Return to the position before the non-whitespace char.
+			reader.Seek(-1, io.SeekCurrent)
+			return nil
+		}
+	}
+	return nil
+}
+
+// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-4.2.3
+func parseParameterisedList(reader *strings.Reader) ([]parameterisedIdentifier, error) {
+	items := []parameterisedIdentifier{}
+	for reader.Len() > 0 {
+		item, err := parseParameterisedIdentifier(reader)
+		if err != nil {
+			return nil, errors.Wrap(err, "parsing parameterised identifier")
+		}
+		items = append(items, *item)
+		// Steps 3 and 4 are swapped per https://github.com/httpwg/http-extensions/pull/667.
+		if reader.Len() == 0 {
+			return items, nil
+		}
+		if err := discardOWS(reader); err != nil {
+			return nil, errors.Wrap(err, "discarding OWS")
+		}
+		comma, err := reader.ReadByte()
+		if err != nil {
+			return nil, errors.Wrap(err, "reading comma")
+		}
+		if comma != ',' {
+			return nil, errors.New("expected comma")
+		}
+		if err := discardOWS(reader); err != nil {
+			return nil, errors.Wrap(err, "discarding OWS")
+		}
+		if reader.Len() == 0 {
+			return nil, errors.New("expected another param-id")
+		}
+	}
+	return nil, errors.New("reached unexpected end of parseParameterisedList")
+}
+
+// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-4.2.4
+func parseParameterisedIdentifier(reader *strings.Reader) (*parameterisedIdentifier, error) {
+	primaryID, err := parseIdentifier(reader)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing primary identifier")
+	}
+	// NOTE: The initial version of AMP-Cache-Transform does not support
+	// any parameters, so we don't bother to implement a parser for them.
+	// No point in distinguishing syntactic failure from semantic failure.
+	// Once we add parameters, we'll need to implement parsers for the
+	// subset item that they use. See Items here:
+	// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-3.4
+	return &parameterisedIdentifier{primaryID}, nil
+}
+
+// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-3.8
+func isIdentifier(c byte) bool {
```

Comment:
naming. isIdentifierByte? isIdentifierChar?

Oddly named. It neither handles the first byte of an identifier (it's too lax, as that can only be lcalpha) nor does it verify the whole identifier.

Reply:
also nit, but maybe put this below the isdigit and islcalpha f'ns, since it depends on them, and places this method closer to it's only dependent.

Reply:
Done and done.